### PR TITLE
[Merged by Bors] - feat: port `Algebra.Hom.Iterate`

### PR DESCRIPTION
**Pull request by winstonyin**

### CODE REVIEW

File: Mathlib/Algebra/Hom/Iterate.lean [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+--what should be the namespace for this section?
+section Monoid
+
+variable [Monoid G] (a : G) (n : ℕ)
+
+@[simp, to_additive]
+theorem smul_iterate [MulAction G H] : ((· • ·) a : H → H)^[n] = (· • ·) (a ^ n) :=
+  funext fun b =>
+    Nat.recOn n (by dsimp only; rw [iterate_zero, id.def, pow_zero, one_smul G b])
+    fun n ih => by dsimp only; rw [iterate_succ', comp_apply, ih, pow_succ, mul_smul]
```
Comment (winstonyin) [inline Mathlib/Algebra/Hom/Iterate.lean:182]: Need `dsimp only` before `rw` works. This is happening quite frequently.

File: Mathlib/Algebra/Hom/Iterate.lean [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+  hom_coe_pow _ rfl (fun _ _ => rfl) _ _
+#align equiv.perm.coe_pow Equiv.Perm.coe_pow
+
+--what should be the namespace for this section?
+section Monoid
+
+variable [Monoid G] (a : G) (n : ℕ)
+
+@[simp, to_additive]
+theorem smul_iterate [MulAction G H] : ((· • ·) a : H → H)^[n] = (· • ·) (a ^ n) :=
```
Comment (eric-wieser) [inline Mathlib/Algebra/Hom/Iterate.lean:179]: `(· • ·) a` is not a faithful port of `(•) a` which means `Smul.smul a`

Comment (winstonyin) [reply Mathlib/Algebra/Hom/Iterate.lean:179]: Would you recommend `(· • ·) a` or `Smul.smul a` during port?

Comment (winstonyin) [reply Mathlib/Algebra/Hom/Iterate.lean]: I've changed them to `(· * a)`

**Comment by kbuzzard** [comment]
I've just added all the #aligns which to_additive doesn't generate.

File: Mathlib/Algebra/Hom/Iterate.lean [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+#align zsmul_iterate zsmul_iterate
+
+end Group
+
+section Semigroup
+
+variable [Semigroup G] {a b c : G}
+
+-- Porting note: need `dsimp only`, see https://leanprover.zulipchat.com/#narrow/stream/
+-- 287929-mathlib4/topic/dsimp.20before.20rw/near/317063489
```
Comment (kbuzzard) [inline Mathlib/Algebra/Hom/Iterate.lean:239]: Yeah, the issue is https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/dsimp.20before.20rw/near/317064195 : `(a * ·) t` expands to `(fun x ↦ a * x) t` in lean 4, but `((*)a) t` expands to `a * t` in lean 3, meaning that sometimes you have to `dsimp only` in Lean 4 to get back on track. You could mention this in the porting note but that's what's going on.

**Review by kbuzzard — APPROVED** [review]
This PR is fine. Thanks!

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**